### ISSUE TITLE
[Tracking] "Replay effect" when sequential actions are taken offline, and user subsequently comes online

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#12775]
### Problem
With Pattern B, offline creation events trigger a "pending" action that clears when the app comes back online. Similarly for deleting or removing content, a "strikethrough" action appears while offline and clears when the app comes back online. The pickle is that if a user takes a creation and deletion action while offline – whether that be uploading/removing an attachment, sending/deleting a message, adding/removing workspace members – content will flash or "replay" because actions in our app happen sequentially. 

Here's an example of how this "replay" happens in practice:
 - User invites a workspace member offline, we show it as pending to create
 - User removes a workspace member offline, we show it as pending to delete
 - When the User comes online, we remove the row optimistically because the last action taken was to delete
 
However, again because of sequentiality, we see flashing in practice. The create action will run first, returning the data and showing the row. Next, the deleted action runs and we remove the row. Here's another pickle: when the creation action is run, we also send an email inviting the workspace member, but then that email is immediately invalidated. Confusing right?

**tldr:** When a user takes multiple write actions affecting the same Onyx key while offline and then goes back online, the actions are “replayed” which causes the UI to update unnecessarily and can cause flickering. We have already tracked 5 issues related to this problem in the App.

### Solution
~~1. https://github.com/Expensify/Expensify/issues/270217 **react-native-onyx:** Create `Onyx.batchUpdate` which will do the same thing as Onyx.update, except that it won't notify subscribers until the last update is applied.~~ Not necessary, Onyx already batches updates.
2. https://github.com/Expensify/Expensify/issues/270219
2.1. **App:** Send the Pusher `socket_id` with all requests.
2.2. **App:** [HOLD 1] While the SequentialQueue is processing, create a list of Onyx updates from each response. Once the SequentialQueue is empty, batch update Onyx.
3. https://github.com/Expensify/Expensify/issues/270220
3.1. **Web-Expensify** Send the Pusher `socket_id` with all Pusher events, [preventing them from being delivered](https://pusher.com/docs/channels/server_api/excluding-event-recipients/) to that client.
3.2. **Web-Expensify** [HOLD 2] Modify our api to return all Onyx updates in the response that were pushed to the requesting accountID.

### Known Issues
 - https://github.com/Expensify/App/issues/12743
 - https://github.com/Expensify/App/issues/12000
 - https://github.com/Expensify/App/issues/13310
 - https://github.com/Expensify/App/issues/15550
 - https://github.com/Expensify/App/issues/15641
 - https://github.com/Expensify/App/issues/16530
 - https://github.com/Expensify/App/issues/13250
 - https://github.com/Expensify/App/issues/17110
 - https://github.com/Expensify/App/issues/17081
 - https://github.com/Expensify/App/issues/16776
 - https://github.com/Expensify/App/issues/17197
 - https://github.com/Expensify/App/issues/16728
 - https://github.com/Expensify/App/issues/18102
 - https://github.com/Expensify/App/issues/12775

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0110088603c45991d5</li>
        <li>Upwork Job ID: 1625975375525715968</li>
        <li>Last Price Increase: 2023-02-15</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by JmillsExpensify** [comment]
I created this issue last week, but it's honestly pretty annoying to have this in the queue given all the immediate work we have in front of us elsewhere. I want to believe that we can come back to this one and remember this issue/convo, though curious for your thoughts.

**Comment by iwiznia** [comment]
It's a tracking issue, why does it bother you that exists? We don't have to work on it now...

**Comment by trjExpensify** [comment]
~Take a seat on this chaise longue, Jason. Talk to me... tell me how it makes you feel?~ 

I'm with Ioni. I think it's fine, let's put the `planning` label on it and circle back to it around N7.

**Comment by JmillsExpensify** [comment]
Nice, I'm good with that!

**Comment by JmillsExpensify** [comment]
Here's [another manifestation](https://expensify.slack.com/archives/C03TQ48KC/p1669639790558289?thread_ts=1669401146.808179&cid=C03TQ48KC) of this convo.

**Comment by JmillsExpensify** [comment]
Still on hold for WAQ.

**Comment by JmillsExpensify** [comment]
Added that one to our OP of known issues.

**Comment by neil-marcellini** [comment]
Thanks! I'm actually going to have that issue closed soon, and the part where the messages are out of order will be fixed in https://github.com/Expensify/App/issues/13310. I'll update the description to link that instead.

**Comment by JmillsExpensify** [comment]
Still on hold and not something we're going to prioritize.

**Comment by trjExpensify** [comment]
We have this held on WAQ, so that should be removed now, but do we want to progress with finding a solution to this atm though?

**Comment by neil-marcellini** [comment]
I vote that we start working on a fix now. It feels like a liability to have a bug in one of our most common offline patterns.

**Comment by trjExpensify** [comment]
I tend to agree we should figure out how we're going to plug this gap to improve the handling of adding/removing something offline. The "flashing" effect feels like a bug, though it is by design I believe, so technically it's still a `New feature` right?

**Comment by iwiznia** [comment]
>  though it is by design I believe, so technically it's still a New feature right?

yeah, kind of by design.

**Comment by trjExpensify** [comment]
Nice let's do it! Neil and Ioni, are you guys looking to collab on it?

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0110088603c45991d5

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor Plus for review of internal employee PR - @aimane-chnaif (`Internal`)

**Comment by iwiznia** [comment]
>  Nice let's do it! Neil and Ioni, are you guys looking to collab on it?

I have no ideas on how to solve this issue. I thought about not updating any UI while we are going through the queue, but that sounds like a bad idea, as it would make the UI unresponsive till the queue is finished.

**Comment by JmillsExpensify** [comment]
Hmm yeah I agree with that. It'd make the app feel like it was frozen when it's not.

**Comment by trjExpensify** [comment]
Yeah, it might feel like there's a big delay in the response after coming back online. Is there no way for us to cancel something in the queue if it's effectively a revert of the same UI element? I.e adding/removing a member, changing the priority mode from most recent/focus/most recent.

**Comment by neil-marcellini** [comment]
I had an idea to ignore the Pusher update from the client that made the request. IIRC the Pusher updates that are sent back cause the problem. We should **confirm** this.

For example if I go offline and flip a toggle on, off, on, then go back online, the Pusher updates will come back with on, off on, resulting in the toggle turning off and back on for me. If we were to ignore those updates because they originated from the same client that is receiving the updates, then the problem would be solved.

If I was simultaneously logged into another device and looking at the same page then I would see the toggle flip on, off, and on when the other device went online, but that makes sense because I'm seeing the actions they took while offline.

What do you guys think of that plan? It might be better to discuss in Slack.

**Comment by JmillsExpensify** [comment]
Wow that's pretty interesting, I like it. I think I agree though, let's confirm with the broader group in Slack.

**Comment by iwiznia** [comment]
I don't think we can do that. The optimistic data is just that, the data we believe the server will return, but the server can (and will) return different data that we need to merge.

**Comment by JmillsExpensify** [comment]
Dream crusher.

**Comment by neil-marcellini** [comment]
> I don't think we can do that. The optimistic data is just that, the data we believe the server will return, but the server can (and will) return different data, data that we need to merge.

That's true, but maybe we can mark specific values to ignore pusher updates for. I'm changing this to a weekly. On Friday I will do some investigation into the known issues and raise a discussion in Slack.

**Comment by neil-marcellini** [comment]
I only had time for one small investigation, which was confirming that if you send two messages quickly on a slow connection then they will appear out of order when the first pusher update comes back. I think this is significant because I don't think we could merge requests for the same command together in the Network queue, since the first request could be in flight when the second request comes in.

On Monday I'll investigate more and create the discussion.

**Comment by neil-marcellini** [comment]
I never got around to creating that discussion, but I did do some more investigation.

For the issue where messages are briefly out of order after going online, I confirmed that it doesn't occur if we don't send the Pusher update. Similarly for adding and removing a workspace member, we won't have a reply effect without the Pusher data. I think we can ignore the Pusher update without any problems as long as there is a read api command on the page where the request is made, which ensures that the data is up to date with the server eventually.

> I don't think we can do that. The optimistic data is just that, the data we believe the server will return, but the server can (and will) return different data, data that we need to merge.

Could you describe a case where the server returns different data that we must be aware of right away? If we still accept updates sent from other users / clients then we will be aware of important changes.

A potential solution would be to generate a NewDot clientID and send it with all api requests. From the API side, send the clientID back in each Pusher event. When NewDot receives Push events from the same clientID it will discard them but it will accept all others. I also have idea idea for ignoring only specific parts of an update, but I'm not sure that's necessary.

I'm hoping to take this on as my next daily issue sometime next week, and when I'm ready I'll start a Slack conversation. Fridays seem to be a bad day for conversations.

**Comment by iwiznia** [comment]
What you say makes sense, but I am still doubtful about it... like the app will be in different states if you did the same actions while online or offline? Kind of odd... maybe is fine, but since this is a global level decision that will apply to all APIs and all features, we need to tread carefully to ensure it is a good idea and in the end not cause more problems than the problem we are trying to solve

**Comment by neil-marcellini** [comment]
I'm making this daily. I'll start to tackle it and raise a discussion on Monday.

**Comment by iwiznia** [comment]
Not overdue

**Comment by JmillsExpensify** [comment]
Looks like this came up very recently [in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1678753304968289?thread_ts=1678029220.184279&cid=C01GTK53T8Q). Let's make sure we're all on the same page in terms of next steps.

**Comment by neil-marcellini** [comment]
I presented a P/S and started a [conversation](https://expensify.slack.com/archives/C03TQ48KC/p1679004542742609) in Slack.

**Comment by neil-marcellini** [comment]
I posted [a summary](https://expensify.slack.com/archives/C03TQ48KC/p1679086458423699?thread_ts=1679004542.742609&cid=C03TQ48KC) of the conversations on Slack and in person today.

**Comment by trjExpensify** [comment]
Nice, thanks again for moving this along!

**Comment by neil-marcellini** [comment]
I think we came to agreement on [this solution](https://expensify.slack.com/archives/C03TQ48KC/p1679438676840909?thread_ts=1679004542.742609&cid=C03TQ48KC).

**Comment by neil-marcellini** [comment]
Marc found about about a Pusher `socket_id` which allows us to not send events to specific clients, so I included that in the solution. I wrote out the solution in order of implementation and I created individual issues for each piece. The implementation order should ensure that no changes take place until the PR for the last issue is deployed.

**Comment by neil-marcellini** [comment]
Marc also pointed out [here](https://expensify.slack.com/archives/C03TQ48KC/p1679499739566429?thread_ts=1679004542.742609&cid=C03TQ48KC) that we don't need Onyx.batchUpdate because Onyx already does batching. I'm updating the solution to remove that step. I'll start implementing the other steps later today.

**Comment by neil-marcellini** [comment]
Other than that I didn't quite get to this today.

**Comment by neil-marcellini** [comment]
Short day again, didn't get to this.

**Comment by trjExpensify** [comment]
@neil-marcellini, how do you feel about putting this on `weekly` until the distance detailed is out the door? IMO, that takes priority over this one.

**Comment by JmillsExpensify** [comment]
Agree that this is less important than distance. I'll go ahead and update to `weekly`.

**Comment by neil-marcellini** [comment]
Ok I'm fine with that

**Comment by JmillsExpensify** [comment]
Oh well that's even better then! I still think it's fine to keep a weekly on it, as that is the minimum expectation during which forward momentum should be shown.

**Comment by roryabraham** [comment]
> Not necessary, Onyx already batches updates.

TBC, I think Onyx batches updates on a per-tick basis. I do not think it will wait for the full sequential queue to clear before performing a batch update

**Comment by roryabraham** [comment]
*Question:* Does this only apply to Pattern B? [The "replay effect" can happen with Pattern A as well](https://expensify.slack.com/archives/C01GTK53T8Q/p1678142233039229?thread_ts=1678029220.184279&cid=C01GTK53T8Q).

**Comment by iwiznia** [comment]
Maybe it is, thought it does not matter for solving this, right?

**Comment by roryabraham** [comment]
Removed `Pattern B` from the title, because it was a bit misleading as this applies to patterns A and B.

**Comment by neil-marcellini** [comment]
👀 I'm going to start off by testing my PRs to see if they work. If so then I'll try to get them up for review.

**Comment by neil-marcellini** [comment]
I verified that my changes fix the replay effect when creating a workspace and renaming it while offline. For some reason I have some CORS errors that seem to be coming from the logging command and I'm really confused about where that's coming from.

I see that I'm hitting [this line](https://github.com/Expensify/Web-Expensify/blob/1dcb48033f3d0227995c32ee249ba4559e44dc47/_errorUtilities.php#L46) on a Log request and that my authToken has expired somehow. I also see that an AuthenticatePusher command has failed and I wonder if that's related. I'm pretty confused and I'll have to investigate more another time.

![image](https://user-images.githubusercontent.com/26260477/229656808-1e6de1c2-9622-431f-a03a-8ebe6eb33250.png)

**Comment by neil-marcellini** [comment]
It turns out I overlooked the fact that some requests do not have an authToken, such as logging. That's very obvious in hindsight.

**Comment by neil-marcellini** [comment]
I discovered that I'm not accessing the Pusher socket_id on the front end properly and I should access when it connects [here](https://github.com/Expensify/App/blob/b703eac7951ba946ba63e3fc83e6fe1b5c64f77d/src/libs/PusherConnectionManager.js#L44-L45) instead. Nothing this so I remember to update it next time.

**Comment by neil-marcellini** [comment]
I'm successfully sending the socket_id from the front end now. I did a good amount of debugging and found several more small mistakes that were causing it to not work. I think it's working well now so next time I need to continue testing and then hopefully it's ready.

**Comment by neil-marcellini** [comment]
I ran a few more successful tests. Creating and deleting a workspace while offline isn't working quite right and I saw one error when signing out and back in so I'll need to sort that out later.

**Comment by hayata-suenaga** [comment]
I added https://github.com/Expensify/App/issues/17081 to OP as the issue seems to be related what being discussed here.

**Comment by neil-marcellini** [comment]
👀 working on this again now. There's been a few more bugs reported and it's getting to be a lot to test, so I'll make sure the main original bugs are tested, and then these others can be retested once this is merged.

**Comment by neil-marcellini** [comment]
I fixed a bug where the `queuedOnyxUpdates` were not properly clearing from Onyx. Next time I need to investigate a bug when signing out and back in. I also thoroughly investigated a what's happening with creating and deleting a workspace while offline, and I left an update [here](https://github.com/Expensify/App/issues/15550#issuecomment-1513933360). TLDR: the create workspace is too large / Onyx.update doesn't always apply the updates as a single batch in order. I think I can move forward with this solution and fix that issue after.

**Comment by neil-marcellini** [comment]
I worked on this a bunch today and got everything cleaned up and ready to go. I need to do a little more testing on the App side for the author checklist, and write the descriptions and all that busy work. I'll get to it tomorrow morning and put these up for review.

**Comment by amyevans** [comment]
Added https://github.com/Expensify/App/issues/16728 to the OP

**Comment by neil-marcellini** [comment]
I forgot to write an update, but I did get this up for review 2 days go, waiting on reviews for the App PR now.

**Comment by sakluger** [comment]
I'm having a hard time finding the PR that will fix this issue - @neil-marcellini do you mind linking to it? I'd like to check the status.

**Comment by neil-marcellini** [comment]
Sure. @sakluger Once this PR https://github.com/Expensify/Web-Expensify/pull/36879 is deployed almost all of the replay effect issues should be fixed. We should retest them and then fix any that need special attention. I know one of them at least will need some more work. The PR I linked is held on the corresponding [App PR](https://github.com/Expensify/App/pull/16545) which was deployed to staging on Friday. Once it hits production, hopefully today, we can take the Web-E PR off of hold and merge it since it's already approved.

**Comment by hayata-suenaga** [comment]
thank you Neil for this fix. I'll test my PR then!

**Comment by hayata-suenaga** [comment]
@rushatgabhane I don't think the Web-E PR is on staging. We'll let you know here on this issue once the PR is deployed to staging

**Comment by neil-marcellini** [comment]
The Web-E PR is on staging.

**Comment by hayata-suenaga** [comment]
cc: @rushatgabhane ⬆️

**Comment by neil-marcellini** [comment]
I went through all of those issues and retested many of them. A good chunk of them were fixed and I closed them, and there's also a good chunk that were not fixed. I'm going to focus on #15550 because I identified that Onyx is not "batch updating" as planned. I think if we can fix that then many of the other issues will be fixed too. The assignees should investigate each remaining issue individually and identify exactly why it's happening.

**Comment by neil-marcellini** [comment]
Tim found out that there was an error in the Pusher version we were using, so setting `socket_id` in the event data did not work as expected to prevent the events from being delivered! Once the PR with the Pusher version update is deployed to staging we should retest the outstanding issues.

**Comment by iwiznia** [comment]
Sorry, what? Can you share more information? What was the bug? What is the version we are updating to? Where's the PR?

**Comment by neil-marcellini** [comment]
Yeah sorry if that was unclear. Part of the solution for the replay effect is to prevent sending Pusher events to the requesting client, which prevents updates made while offline from being replayed in the UI. To do that we send the Pusher `socket_id` in the [Pusher event data](https://github.com/Expensify/Web-Expensify/blob/1eb49419b177b91438a9f0be1591d759b48f2862/lib/Push/Service.php#L241-L243), and that follows the [instructions in the documentation](https://pusher.com/docs/channels/server_api/excluding-event-recipients/).

[Tim found out](https://expensify.slack.com/archives/C03TQ48KC/p1684935269590919?thread_ts=1684927247.651649&cid=C03TQ48KC) that pusher-http-php library version 7.2.2 did not properly use the `socket_id` and version 7.2.3 fixes it. That version bump is included in this PR [[Web-Expensify] Add a parameter to PHP add/remove emoji reaction commands for useEmojiReactions](https://github.com/Expensify/Web-Expensify/pull/37193). If we need this to move faster and that PR isn't merged soon we could extract the change to a separate PR, but I think it will merge soon and there's no rush.

**Comment by neil-marcellini** [comment]
https://github.com/Expensify/Web-Expensify/pull/37193 is now on staging with the fix to properly prevent sending Push events to the requesting client. Tomorrow I will retest some of the remaining replay effect bugs.

**Comment by Christinadobrzyn** [comment]
Hey team! Could you double-check if this GH - https://github.com/Expensify/App/issues/19651#issuecomment-1574304811 will be resolved with this fix?

**Comment by iwiznia** [comment]
I don't think it will

**Comment by neil-marcellini** [comment]
Unfortunately the update to fix Pusher and make it correctly prevent sending updates based on the `socket_id` was [reverted](https://github.com/Expensify/Web-Expensify/pull/37696#issuecomment-1577568299). So that means the fix for this issue is still not live.

I believe it was reverted due to the other changes in that PR. I'll double check on that and then I'll put up a quick PR which only updates the pusher-php-server version to 7.2.3. I will also request to CP it to staging.

**Comment by neil-marcellini** [comment]
The fix is now on staging so we can take related issues off of HOLD and retest. If there are still problems we'll have to debug and figure out why.

I know there is still one problem replay effect problem identified in #15550 which still needs to be fixed. The problem is that sometimes calling Onyx.update with a list of updates doesn't produce the desired end result. An example of the problem is shown [here](https://github.com/Expensify/App/issues/15550#issuecomment-1513933360). Please only HOLD issues on this one if you identify that we're facing this exact problem, or some other general replay problem that is unsolved.

**Comment by neil-marcellini** [comment]
Tomorrow I'll work on one or two of my weekly issues, this one has lower priority.

**Comment by neil-marcellini** [comment]
I created this Onyx issue https://github.com/Expensify/react-native-onyx/issues/266 which identifies the root problem for some of these lingering issues. If you are investigating a remaining replay effect issue please check if it shares this root cause and add a link on the Onyx issue if so. Also feel free to jump in on the conversation and add your 2 cents if you think it's important to fix.

**Comment by neil-marcellini** [comment]
I still need to make a proposal for that Onyx issue, haven't gotten to that yet.

**Comment by trjExpensify** [comment]
Any movement here, @neil-marcellini?

**Comment by trjExpensify** [comment]
Fair enough!

**Comment by neil-marcellini** [comment]
Still putting this off since it's lower priority

**Comment by neil-marcellini** [comment]
Yep

**Comment by neil-marcellini** [comment]
I've been OOO and prioritizing distance requests.

**Comment by trjExpensify** [comment]
Suspect finishing up distance requests is still taking priority here before circling back to close this one out, right @neil-marcellini?

**Comment by neil-marcellini** [comment]
Yep

**Comment by neil-marcellini** [comment]
I still haven't had time for this.

**Comment by JmillsExpensify** [comment]
Still in the queue

**Comment by JmillsExpensify** [comment]
Same as above.

**Comment by neil-marcellini** [comment]
Bumping down to monthly. Or should we just close it and re-open if we decide it's important again?

**Comment by trjExpensify** [comment]
I don't think we should close it, because it is a pretty weird quirk of our offline UX.

**Comment by neil-marcellini** [comment]
@trjExpensify @aimane-chnaif is requesting compensation for reviewing [#16545](https://github.com/Expensify/App/pull/16545). Could you please handle that when you have time?

**Comment by trjExpensify** [comment]
Oh, that's old haha. @aimane-chnaif that would be a straight $1000 for the internal PR review, correct?

**Comment by neil-marcellini** [comment]
> @neil-marcellini will confirm the amount

Unfortunately @aimane-chnaif never completed the checklist so I don't think he should be paid for that review.

**Comment by neil-marcellini** [comment]
👀 Ok I'm coming back to this for the first time in a while. When I last worked on it, long ago, it was working except for one tiny issue in Onyx.

I'm going to see if the Onyx issue and some of the linked issues are still reproducible. If so then I will recruit an engineer from Margelo to work their Onyx magic and get this wrapped up by the end of the year. I'll stick around to review proposals and PRs.

**Comment by neil-marcellini** [comment]
I verified that the issues are still reproducible. Most of the action will be happening in that Onyx issue because I think it's the root of the problem, so please follow along there if you like.

**Comment by JmillsExpensify** [comment]
Perhaps for clarity, is Margelo going to pick this up?

**Comment by JmillsExpensify** [comment]
Bump on that last question.

**Comment by chrispader** [comment]
Yes, i've started working on this! 👍 

(Sorry for the delay, wanted to start earlier but got busy with theme switching)

**Comment by chrispader** [comment]
This PR should fix the problem by fixing the `Onyx.update` order problem: https://github.com/Expensify/react-native-onyx/issues/266#top

https://github.com/Expensify/react-native-onyx/pull/437

**Comment by chrispader** [comment]
Testing the changes tmrw!

**Comment by chrispader** [comment]
Onyx bump PR created here: https://github.com/Expensify/App/pull/33554

**Comment by neil-marcellini** [comment]
❗ 🔥 Heads up, I'm going to be OOO working 0% from 1/1-1/12/24. I'll be back on 1/15/24. I'm planning to work 50% tomorrow 12/29. Please feel free to un-assign, re-assign, whatever. When I return I'll pick back up whatever I can and do my best to GSD in order of priority. I will also post all of my assigned issues in #engineering-chat in Slack to try to recruit volunteers.

**Comment by roryabraham** [comment]
@chrispader as you know the Onyx bump was reverted. Since we need to repeat that Onyx bump PR, can you please upgrade the Node version in react-native-onyx to 20.10.0 and npm to 10.2.3, then do the same in E/App?

**Comment by chrispader** [comment]
> @chrispader as you know the Onyx bump was reverted. Since we need to repeat that Onyx bump PR, can you please upgrade the Node version in react-native-onyx to 20.10.0 and npm to 10.2.3, then do the same in E/App?

yes, i can do that 👍 @roryabraham

**Comment by roryabraham** [comment]
Merged the Node bump in Onyx

**Comment by trjExpensify** [comment]
A tad out of the loop here, is payment actually due?

**Comment by trjExpensify** [comment]
Dang, an issue from January? 

Payment summary as follows: 

- $500* to @aimane-chnaif for the C+ review

*$500 because this is an issue created prior to the switch to $250. Offer sent.

**Comment by trjExpensify** [comment]
Paid, closing!
